File: 19-remove-nth-node-from-end-of-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
    pre := &ListNode{}
    pre.Next = head
    header := pre
    fast := head
    for n > 1 {
        fast = fast.Next
        n -= 1
    }
    for fast.Next != nil {
        fast = fast.Next
        pre = pre.Next
    }
    pre.Next = pre.Next.Next
    return header.Next
}

File: bwu/linked_list/19-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        
        slow, fast = dummy, dummy
        for i in xrange(n):
            fast = fast.next
            
        while fast.next:
            fast = fast.next
            slow = slow.next
            
        slow.next = slow.next.next
        return dummy.next
